File: Python/SOLID/solid_1.py
"""
SOLID principles
1: Single-responsibility principle
"""


class Journal:
    """
    Journal class
    """

    def __init__(self) -> None:
        """
        Initialize
        """
        self.entries = []
        self.count = 0

    def add_entry(self, text: str) -> None:
        """
        Add entry to the journal
        """
        self.count += 1
        self.entries.append(f"{self.count}: {text}")

    def remove_entry(self, pos: int) -> None:
        """
        Remove entry from the journal
        """
        del self.entries[pos]
        self.count -= 1

    def __str__(self) -> str:
        """
        Override print method
        """
        return "\n".join(self.entries)

    # This is where it breaks the principle
    # def save(self, filename) -> None:
    #     """
    #     Save journal
    #     """
    #     file = open(filename, "w", encoding="")
    #     file.write(str(self))
    #     file.close()

    # def load(self, filename) -> None:
    #     """
    #     Load from file
    #     """

    # def load_from_web(self, uri) -> None:
    #     """
    #     Load from web address
    #     """


# Solution: create new class to handle new functionality
class PersistanceManager:
    """
    Manage the Journal
    """

    @staticmethod
    def save_to_file(journal: Journal, filename: str) -> None:
        """
        Save journal to file
        """
        file = open(filename, "w", encoding="utf8")
        file.write(str(journal))
        file.close()


if __name__ == "__main__":
    new_journal = Journal()
    new_journal.add_entry("This is the first entry.")
    new_journal.add_entry("Second day is also boring.")
    print(f"Journal entries: \n{new_journal}")

    PATH_TO_FILE = r"./journal.txt"
    PersistanceManager.save_to_file(new_journal, PATH_TO_FILE)

File: Go/SOLID/solid_1.go
// SOLID principles
// 1: Single-responsibility principle

package main

import (
	"fmt"
	"io/ioutil"
	"net/url"
	"strings"
)

var entryCount = 0

type Journal struct {
	entries []string
}

func (j *Journal) AddEntry(text string) int {
	entryCount++
	entry := fmt.Sprintf("%d: %s", entryCount, text)
	j.entries = append(j.entries, entry)

	return entryCount
}

func (j *Journal) RemoveEntry(index int) {
	j.entries = append(j.entries[:index], j.entries[index+1:]...)
}

// separation of concerns
// God Object = antipattern

func (j *Journal) String() string {
	return strings.Join(j.entries, "\n")
}

func (j *Journal) Save(filename string) {
	_ = ioutil.WriteFile(filename, []byte(j.String()), 0644)
}

func (j *Journal) Load(filename string) {
	// ...
}

func (j *Journal) LoadFromWeb(url *url.URL) {
	// ...
}

// solution
var LineSeparator = "\n"

func SaveToFile(j *Journal, filename string) {
	_ = ioutil.WriteFile(filename, []byte(strings.Join(j.entries, LineSeparator)), 0644)
}

// even better
type Persistence struct {
	lineSeparator string
}

func (p *Persistence) SaveToFile(j *Journal, filename string) {
	_ = ioutil.WriteFile(filename, []byte(strings.Join(j.entries, p.lineSeparator)), 0644)
}

func main() {
	j := Journal{}
	j.AddEntry("This is the first entry.")
	j.AddEntry("Second day is also boring.")
	fmt.Println(j.String())

	// SaveToFile(&j, "journal.txt")

	p := Persistence{"/r/n"}
	p.SaveToFile(&j, "journal.txt")
}
